Update c curl generator to support float

amoweb opened the pull request:
<!-- Enter details of the change here. Include additional tests that have been done, reference to the issue for tracking, etc. -->

Fix for issue #21092

<!-- Please check the completed items below -->
### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [ x Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package || exit
  ./bin/generate-samples.sh ./bin/configs/*.yaml || exit
  ./bin/utils/export_docs_generators.sh || exit
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming `7.x.0` minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [ ] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.


@eafer

eafer commented:
Hi! I just took a quick look at the new generated samples, and I see that petId for deletePet becomes a pointer. What's the point of that? The petId is not nullable, right?

eafer commented:
Also, would you mind adding some tests for these issues to the `petstore.yaml` for C? That would make it much easier to understand what you are doing, and it would prevent other people from breaking this again in the future.

amoweb commented:
Hi! Thanks for your answer! Considering your comments, I changed the strategy. I added an example in petstore.yaml and I no longer change the function signature.

eafer commented:
Thanks. I can see a new `UserAPI_testInt32Int64FloatDouble()` function in the generated samples, but I'm not seeing the changes to the yaml file that made that happen. Am I missing something, or maybe you forgot to commit that?

amoweb commented:
Sorry. I commited the missing yaml file.

eafer commented:
Thanks. Why did you set `MAX_NUMBER_LENGTH_FLOAT` to 32, is this an arbitrary limit? For big float numbers the output of `snprintf()` will be off. I think in this case it would be better to call snprintf twice, the first time to get the size of the buffer needed.

eafer commented:
> I defined a default size and reallocate only if needed (so in most case, we only call snprintf once).

What's the advantage of this over just calling snprintf() twice? My concern is that the long float path will almost never get tested.

> I think %.7e for float and %.16e for double is more relevant since they roughly correspond to the float precision

Ah yes, that makes more sense.

eafer commented:
By the way, what's this change about? Are you using the returned value anywhere?

```
+        int s = snprintf(valueQuery_{{{paramName}}}, MAX_NUMBER_LENGTH, "%d", *{{{paramName}}});
```

amoweb commented:
I think it's less performant to call snprintf() twice. The initial size is sufficient for most numbers (except very big ones). I can still change it if you prefer.

I removed the useless "int s = ".

eafer commented:
> I think it's less performant to call snprintf() twice.

Sure, but this code only runs right before making a blocking network request, so I don't think it could ever be a bottleneck. I think it's better to keep it simple (unless you have any benchmarks that show otherwise).

> I removed the useless "int s = ".

Thanks.

eafer commented:
Thank you for making the changes. This looks good to me now @wing328

amoweb commented:
@eafer @wing328 Hello, did you had time to have a look to my PR? Thanks

wing328 commented:
Thanks for the PR, which has been merged.

Sorry for the delay as there are too many PRs for this project.

Have a good day.
